torch/csrc/jit/passes/constant_propagation.cpp

```diff
@@ -119,22 +120,40 @@ void inlineIf(Node* n, const AliasDb& aliasDb) {
   inlineIfBody(n->blocks().at(block_index));
 }
 
+void replaceAndRemoveIfOutput(Node* n, size_t i, Value* replacement) {
+  n->outputs().at(i)->replaceAllUsesWith(replacement);
+  n->eraseOutput(i);
+  n->blocks().at(0)->eraseOutput(i);
+  n->blocks().at(1)->eraseOutput(i);
+}
+
 // remove extra outputs from the node
 bool removeExtraIfOutputs(Node* n) {
   AT_CHECK(n->kind() == prim::If, "Only supported for If nodes");
   auto true_block = n->blocks()[0];
   auto false_block = n->blocks()[1];
+  auto graph = n->owningGraph();
   auto initial_outputs = true_block->outputs().size();
+  WithInsertPoint guard(n);
   for (size_t i = 0; i < true_block->outputs().size();) {
+    auto t_out = true_block->outputs().at(i);
+    auto f_out = false_block->outputs().at(i);
+
     // neither block changes the output value
     if (true_block->outputs()[i] == false_block->outputs()[i]) {
-      n->outputs().at(i)->replaceAllUsesWith(true_block->outputs()[i]);
-      n->eraseOutput(i);
-      true_block->eraseOutput(i);
-      false_block->eraseOutput(i);
-    } else {
-      i++; // increment bc we didn't remove current index
+      replaceAndRemoveIfOutput(n, i, true_block->outputs()[i]);
+      continue;
+    }
+
+    // true block output is constant and constant matches false block output
+    auto maybe_const = toIValue(t_out);
+    if (maybe_const && t_out->node() == f_out->node()) {
```

Comment:
having trouble understanding this one. If `t_out->node() == f_out->node()`, doesn't that mean that the node precedes the if statement (since it has to be accessible to both blocks)? If so, why did we ever emit if/else block outputs for it in the first place, and why do we have to insert a new constant, when presumably one already exists?

Reply:
Agreed this is not very readable. I'm using the overloaded `==` from node_hashing, which checks that the Attributes are equal and that the output has the same type. I'll rewrite with method call. It doesn't mean that the node precedes the if statement. A constant node with the same attribute in each block is an example of when you would have to insert a new constant node before the if node.

Reply:
Good catch - I was trying to use the Equal Node `==` operator but dont think I actually was.